File: server/src/main/java/net/progruzovik/dissent/captain/AiCaptain.java
package net.progruzovik.dissent.captain;

import net.progruzovik.dissent.battle.model.Unit;
import net.progruzovik.dissent.battle.model.field.gun.GunCells;
import net.progruzovik.dissent.battle.model.util.Cell;
import net.progruzovik.dissent.model.entity.GunEntity;
import net.progruzovik.dissent.model.entity.HullEntity;
import net.progruzovik.dissent.model.domain.Ship;
import net.progruzovik.dissent.model.event.Event;
import net.progruzovik.dissent.model.event.EventSubject;
import net.progruzovik.dissent.repository.GunRepository;
import net.progruzovik.dissent.repository.HullRepository;
import org.springframework.context.annotation.Scope;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Random;

@Component
@Scope("prototype")
public final class AiCaptain extends AbstractCaptain {

    public AiCaptain(HullRepository hullRepository, GunRepository gunRepository) {
        HullEntity aiHull = hullRepository.findById(6).get();
        GunEntity artillery = gunRepository.findById(2).get();
        getShips().add(new Ship(aiHull, artillery, null));
        getShips().add(new Ship(hullRepository.findById(4).get(), gunRepository.findById(3).get(), null));
        getShips().add(new Ship(aiHull, artillery, null));
    }

    @NonNull
    @Override
    public String getId() {
        return "AI_CAPTAIN";
    }

    @Override
    public void accept(Event<?> event) {
        if (event.getSubject().equals(EventSubject.NEW_TURN_START)
                && getBattle() != null && getBattle().isIdBelongsToCurrentCaptain(getId())) {
            Unit unit = getBattle().getCurrentUnit();
            if (unit.getShip().getFirstGun() != null) {
                boolean canCurrentUnitMove = true;
                while (unit.getActionPoints() >= unit.getShip().getFirstGun().getShotCost()
                        && canCurrentUnitMove) {
                    int firstGunId = unit.getShip().getFirstGun().getId();
                    GunCells gunCells = getBattle().getGunCells(firstGunId);
                    if (gunCells.getTargets().isEmpty()) {
                        List<Cell> reachableCells = getBattle().getReachableCells();
                        if (reachableCells.isEmpty()) {
                            canCurrentUnitMove = false;
                        } else {
                            Random random = new Random();
                            getBattle().moveCurrentUnit(getId(), reachableCells.get(random.nextInt(reachableCells.size())));
                        }
                    } else {
                        getBattle().shootWithCurrentUnit(getId(), firstGunId, gunCells.getTargets().get(0).getCell());
                    }
                }
            }
            endTurn();
        }
    }
}

File: server/src/main/kotlin/net/progruzovik/dissent/captain/AiCaptain.kt
package net.progruzovik.dissent.captain

import net.progruzovik.dissent.model.domain.Ship
import net.progruzovik.dissent.model.domain.battle.Battle
import net.progruzovik.dissent.model.entity.GunEntity
import net.progruzovik.dissent.model.entity.HullEntity
import net.progruzovik.dissent.model.event.Event
import net.progruzovik.dissent.model.event.EventName
import net.progruzovik.dissent.repository.GunRepository
import net.progruzovik.dissent.repository.HullRepository
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class AiCaptain(hullRepository: HullRepository, gunRepository: GunRepository) : AbstractCaptain() {

    override val id = "AI_CAPTAIN_ID"

    init {
        val baseHull: HullEntity = hullRepository.findById(6).get()
        val artillery: GunEntity = gunRepository.findById(2).get()
        ships.add(Ship(baseHull, artillery, null))
        ships.add(Ship(hullRepository.findById(4).get(), gunRepository.findById(3).get(), null))
        ships.add(Ship(baseHull, artillery, null))
    }

    override fun accept(event: Event<*>) {
        val battle: Battle = this.battle ?: return
        if (event.name == EventName.NEW_TURN_START && battle.isIdBelongsToCurrentCaptain(id)) {
            if (battle.currentUnit.ship.firstGun != null) {
                var canCurrentUnitMove = true
                while (battle.currentUnit.actionPoints >= battle.currentUnit.ship.firstGun!!.shotCost && canCurrentUnitMove) {
                    val firstGunId = battle.currentUnit.ship.firstGun!!.id
                    val gunCells = battle.getGunCells(firstGunId)
                    if (gunCells.targets.isEmpty()) {
                        val reachableCells = battle.reachableCells
                        if (reachableCells.isEmpty()) {
                            canCurrentUnitMove = false
                        } else {
                            battle.moveCurrentUnit(id, reachableCells[(0 until reachableCells.size).random()])
                        }
                    } else {
                        battle.shootWithCurrentUnit(id, firstGunId, gunCells.targets[0].cell)
                    }
                }
            }
            endTurn()
        }
    }
}
